csharp/ql/src/semmle/code/csharp/dataflow/SSA.qll

```diff
@@ -223,34 +275,23 @@ module Ssa {
       outRefExitRead(bb, i, v, node) and
       rk = OutRefExitRead()
       or
-      capturedVarExitRead(bb, i, v, node) and
-      rk = CapturedVarExitRead()
-      or
       refReadBeforeWrite(bb, i, v, node) and
       rk = RefReadBeforeWrite()
     }
 
     private predicate outRefExitRead(
-      ControlFlow::BasicBlocks::ExitBlock ebb, int i, LocalScopeSourceVariable v,
-      ControlFlow::Nodes::ExitNode node
+      BasicBlock bb, int i, LocalScopeSourceVariable v, ControlFlow::Nodes::AnnotatedExitNode node
     ) {
+      node.isNormal() and
       exists(LocalScopeVariable lsv |
         lsv = v.getAssignable() and
-        ebb.getNode(i) = node and
+        bb.getNode(i) = node and
         node.getCallable() = lsv.getCallable()
       |
-        lsv.isRef() or lsv.(Parameter).isOut()
-      )
-    }
-
-    private predicate capturedVarExitRead(
-      ControlFlow::BasicBlocks::ExitBlock ebb, int i, LocalScopeSourceVariable v,
-      ControlFlow::Nodes::ExitNode node
-    ) {
-      exists(BasicBlock bb | variableDefinition(bb, _, v, _) |
-        ebb.getNode(i) = node and
-        bb.getCallable() = ebb.getCallable() and
-        bb.getCallable() != v.getAssignable().getCallable()
+        lsv.(Parameter).isOutOrRef()
+        or
+        lsv.isRef() and
+        strictcount(v.getAnAccess()) > 1
```

Comment:
Yes, it is for `ref` locals. The check is there to avoid SSA definitions for cases like this: https://github.com/github/codeql/blob/main/csharp/ql/test/library-tests/dataflow/ssa/Consistency.cs#L55. Before the refactoring, this was taken care of by https://github.com/github/codeql/pull/4923/files#diff-7a5016d3998d3aa9a71ee689d871333e9bffa227aae862f5d5183a131fdc4f40L1800-L1810, but since we are now using a common liveness analysis, I applied this (less precise) check instead.